### PR TITLE
[WIP] Porting ros2

### PR DESCRIPTION
**Pull request by hsd-dev**
Originally from https://github.com/PepperlFuchs/pf_lidar_ros_driver/pull/82 started by @wawanbreton

- [x] CI: https://github.com/PepperlFuchs/pf_lidar_ros_driver/pull/83/commits/de9e3adfc232b766271b7c5a3e57a3477c90528d, https://github.com/PepperlFuchs/pf_lidar_ros_driver/pull/83/commits/7d354f603cf16a2f3389d20c122cb16a3f41dd2c
- [x] README: https://github.com/PepperlFuchs/pf_lidar_ros_driver/pull/83/commits/8d3a2f918bda80feaf5ec71001f82735d4efd633
- [x] pf_interfaces
  - [x] package.xml and CMakeLists.txt
- [x] pf_description
  - [x] package.xml and ~~CMakeLists.txt~~ setup.py: https://github.com/PepperlFuchs/pf_lidar_ros_driver/pull/83/commits/9476ad63b870801110c965d7b480135093a04f58
  - [x] launch files: https://github.com/PepperlFuchs/pf_lidar_ros_driver/pull/83/commits/64c4d08a72eaa3a438f659ed5c97431c0ac8c611
  - [x] update rviz config: https://github.com/PepperlFuchs/pf_lidar_ros_driver/pull/83/commits/d01c6d13dc8092307f9531238643dd0afe73d084
- [X] pf_driver
  - [x] package.xml and CMakeLists.txt: https://github.com/PepperlFuchs/pf_lidar_ros_driver/pull/83/commits/0248e7493d6562fe3ebc6dc52c860437c7be5634
  - [x] Node, logging, interfaces, callbacks: https://github.com/PepperlFuchs/pf_lidar_ros_driver/pull/83/commits/0248e7493d6562fe3ebc6dc52c860437c7be5634 
  - [x] launch files
  - [x] tests
  - [x] cpp code refactoring to improve parameter management: https://github.com/PepperlFuchs/pf_lidar_ros_driver/commit/1244f3d42f1cd3584cb84fd5127f365a69f4f52f (currently removed since https://github.com/PepperlFuchs/pf_lidar_ros_driver/pull/77 and https://github.com/PepperlFuchs/pf_lidar_ros_driver/pull/81 changes quite some cpp code)

### CODE REVIEW

**Comment by wawanbreton** [comment]
I got permission denied when pushing to this branch, can you give me some rights ?

File: pf_driver/include/pf_driver/pf/r2000/pfsdp_2000.hpp [hunk 1]
```diff
@@ -28,119 +28,46 @@ class PFSDP_2000 : public PFSDPBase
...
-      config_->watchdogtimeout = config.watchdogtimeout;
-    }
-    else if (level == 22)
-    {
-      config_->start_angle = config.start_angle;
-    }
-    else if (level == 23)
-    {
-      config_->max_num_points_scan = config.max_num_points_scan;
+    for(const auto &parameter : parameters)
```
Comment (ipa-hsd) [inline pf_driver/include/pf_driver/pf/r2000/pfsdp_2000.hpp:88]: nice :)

Comment (wawanbreton) [reply pf_driver/include/pf_driver/pf/r2000/pfsdp_2000.hpp:88]: Dynamic parameters are indeed much simpler in ROS2 !

**Comment by ipa-hsd** [comment]
The current code seems to be valid only for `foxy`. We might need to have separate branches for `galactic` and upwards https://github.com/PepperlFuchs/pf_lidar_ros_driver/runs/6810174179?check_suite_focus=true#step:4:685

**Comment by wawanbreton** [comment]
> The current code seems to be valid only for `foxy`. We might need to have separate branch for `galactic` and upwards 

What makes you say that ? I'm working with galaxy only

**Comment by ipa-hsd** [comment]
@wawanbreton please check the CI output linked above. Didn't you get that error?

**Comment by wawanbreton** [comment]
No I didn't get the error, pcl-conversion is actually available for galactic

**Comment by ipa-hsd** [comment]
right, we have commented out the deps https://github.com/PepperlFuchs/pf_lidar_ros_driver/blob/porting-ros2/pf_driver/package.xml#L13. Since CI starts with a clean docker, these packages have not been installed.

**Comment by ipa-hsd** [comment]
OK the CI seems to be working now: https://github.com/PepperlFuchs/pf_lidar_ros_driver/runs/6811233956?check_suite_focus=true#step:4:1582

**Comment by wawanbreton** [comment]
My commit containing the DAE files instead of the STL has been wiped out, was it intentional ?

**Comment by ipa-hsd** [comment]
Also, take a look at the CI output linked above
```
/root/target_ws/src/pf_lidar_ros_driver/pf_driver/src/pf/pf_interface.cpp:220:25: error: missing template arguments before ‘feed_time’
    220 |   std::chrono::duration feed_time = std::min(duration, std::chrono::milliseconds(std::chrono::seconds(60)));
        |                         ^~~~~~~~~
  /root/target_ws/src/pf_lidar_ros_driver/pf_driver/src/pf/pf_interface.cpp:221:46: error: ‘feed_time’ was not declared in this scope
    221 |   watchdog_timer_ = node_->create_wall_timer(feed_time, std::bind(&PFInterface::feed_watchdog, this));
```

**Comment by wawanbreton** [comment]
Commit restored. Please don't force-push unless you are alone on a repos :)

**Comment by wawanbreton** [comment]
Build error should be fixed, I used a declaration that is only available with C++17 but foxy targets C++14

**Comment by ipa-hsd** [comment]
Should we remove these two commits for now? https://github.com/PepperlFuchs/pf_lidar_ros_driver/pull/83/commits/3f1385c7fd3d02e49c5014705406aa031fbfd073 and https://github.com/PepperlFuchs/pf_lidar_ros_driver/pull/83/commits/c6c290fb428674dd6ca729039dc04990b9546594. For this PR I want to focus on porting existing functionality to ROS2 and not add new features mid-way. I can create a new PR with those 2 commits stacks behind this one.

**Comment by wawanbreton** [comment]
I completely agree, I was wondering about splitting them to an other PR. Maybe this should be the case for https://github.com/PepperlFuchs/pf_lidar_ros_driver/pull/83/commits/d75100945368923990de207e8ce2ad479ae97bdf too ?

**Comment by wawanbreton** [comment]
Hi @ipa-hsd, I'm starting to rework the parameters stack because some things are not properly done at initialization. I'm wondering about the parameters philosophy, I can see two possible options :

1. All parameters given to driver are set to the LIDAR, other parameters are forced to default values. LIDAR previous configuration is not taken care of.
2. All parameters given to driver are set to the LIDAR, other parameters are read from the LIDAR and stored into actual parameter values.

Option 1 is more repeatable because whatever you did before, the LIDAR will always start in the same configuration. Option 2 considers you may want to store some configuration into the LIDAR definitely. Personnally, I prefer option 1 because we can easily set all the required parameters into a config file, so I don't really see a need to store configuration on long-term. What do you think should be implemented ?

**Comment by ipa-hsd** [comment]
Thanks for taking a look into this. 

>  some things are not properly done at initialization

Anything specific?

Maybe @ptruka has something to say about this based on the customer feedback? 

@wawanbreton also I am starting to think about the fact the base cpp code (not ROS specific parts) is starting to diverge from the code for ROS1. I am not sure to what extent it is possible to reuse the same classes, but something we might want to keep in mind. One possibility could be whatever inconsistencies / bugs you have fixed in this PR, we could propagate it back to the ROS1 code as well.

**Comment by wawanbreton** [comment]
> Anything specific?

Yesterday I implemented start_angle parameter, but then figured out that it was not applied at startup. I'm working on it.

> One possibility could be whatever inconsistencies / bugs you have fixed in this PR, we could propagate it back to the ROS1 code as well

Until now, I only fixed the bugs I created by changing the parameters description. The management is very different from the one with ROS1, so I'm not sure we could have the same code base for both. I'd be happy to really refactor this part, but I would understand you want to keep consistency between versions. However, I only changed the ROS-related code but all low-level communication code is still the same.

**Comment by ipa-hsd** [comment]
As long as we ensure that the behaviour is the same in ROS1 and ROS2, it should be fine.

**Comment by wawanbreton** [comment]
> As long as we ensure that the behaviour is the same in ROS1 and ROS2, it should be fine. I can think of some tests for that.

Great, I'll keep going this way. And It would make me feel better to have some tests indeed :)

File: pf_driver/src/pf/pfsdp_protocol.cpp [hunk 2]
```diff
@@ -204,6 +203,12 @@ void PFSDPBase::declare_critical_parameters()
   descriptorPort.name = "Initial IP port";
   descriptorPort.description = "See address";
   node_->declare_parameter<int>("port", 0, descriptorPort);
+
+  rcl_interfaces::msg::ParameterDescriptor descriptorTransport;
+  descriptorTransport.name = "IP transport port";
+  descriptorTransport.description = "Can be either tcp or udp";
+  descriptorTransport.read_only = true;
+  node_->declare_parameter<std::string>("transport", "tcp", descriptorTransport);
```
Comment (ipa-hsd) [inline pf_driver/src/pf/pfsdp_protocol.cpp]: Also I didn't notice this earlier, but it's better to separate the protocol-specific code from ROS-specific code. The protocol-specific code is intended to be used like a pure cpp library with no ROS dependency, apart from logging.

**Comment by wawanbreton** [comment]
> it would make sense to use this PR to make only ROS-specific changes and make a new PR for cpp code improvements.

OK, I didn't understand this. The thing is, proper management of parameters will require some architectural changes, so I thought it would be the right time. But I can do something a bit less intrusive yet, even if it is a bit dirty, and make it clean later, if that's what you want.

> The protocol-specific code is intended to be used like a pure cpp library with no ROS dependency, apart from logging

Noted, it perfectly makes sense !

**Comment by ipa-hsd** [comment]
> The thing is, proper management of parameters will require some architectural changes, so I thought it would be the right time.

I agree, but some logic has changed and couple of new features have been introduced in those PRs. I have added that commit as part of TODO in this PR's description.

**Comment by wawanbreton** [comment]
Ok, I will wait for your go to work on this.

**Comment by ipa-hsd** [comment]
The CI is failing because the code is not formatted properly https://github.com/PepperlFuchs/pf_lidar_ros_driver/runs/7018472659?check_suite_focus=true#step:4:174
Could you apply the clang formatting? (If you are using VS Code, right click on the file and "Format document with..."). 

The link above shows which documents need to be updated.

**Comment by wawanbreton** [comment]
I'm not using VS Code, but I tried applying formatting using clang-format. However, the whole file is changed and becomes very different... Is there a formatting profile you are implicitely using ?

**Comment by ipa-hsd** [comment]
This one: https://github.com/PepperlFuchs/pf_lidar_ros_driver/blob/main/.clang-format

**Comment by wawanbreton** [comment]
Seems I got it right :) I configured clang-format to style "FIle" so it probably found it by itself. Thank you for pointing it to me, I'd like to use it for my own projects !

**Comment by ipa-hsd** [comment]
@wawanbreton both https://github.com/PepperlFuchs/pf_lidar_ros_driver/pull/77 and https://github.com/PepperlFuchs/pf_lidar_ros_driver/pull/81 have been merged now. Could you apply https://github.com/PepperlFuchs/pf_lidar_ros_driver/commit/1244f3d42f1cd3584cb84fd5127f365a69f4f52f to the latest `main` branch? Then we can rebase this PR.

**Comment by ipa-hsd** [comment]
No worries, I can make those changes based on https://github.com/PepperlFuchs/pf_lidar_ros_driver/commit/1244f3d42f1cd3584cb84fd5127f365a69f4f52f and you can carry on from there.

**Comment by ipa-hsd** [comment]
@wawanbreton I have rebased this PR from https://github.com/PepperlFuchs/pf_lidar_ros_driver/pull/87. Now we are good to go.

**Comment by ipa-hsd** [comment]
@chenwu-cc thanks for testing the branch. I thought the problem was only on my PC. I will test it today again and get back to you.

**Comment by chenwu-cc** [comment]
> 

May I ask if I use the galactic version could I run porting ros2.

**Comment by ipa-hsd** [comment]
@chenwu-cc could you please try this PR again? I fixed the issue.

**Comment by ptruka** [comment]
Many thanks to everyone for their support!
Due to the general changes to the code and the resulting differences to the ros1 driver, a repository for ros2 was created:
https://github.com/PepperlFuchs/pf_lidar_ros2_driver

**Pull request closed without merging**